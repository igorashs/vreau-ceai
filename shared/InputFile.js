import styled from 'styled-components';
import { Label } from '@/shared/Label';
import { useState } from 'react';

const Wrapper = styled.div`
  display: flex;
  flex-direction: column;
  gap: 7px;
`;

const StyledInputFile = styled.label`
  align-self: flex-start;
  padding: calc(var(--baseline) / 4) calc(var(--baseline) / 2);
  border-radius: 4px;
  border: 0;
  color: var(--text-light);
  background-color: var(--accent-dark);
  box-shadow: 1px 1px 2px #00000066;
  text-transform: lowercase;
  text-decoration: none;

  cursor: pointer;

  :hover {
    background-color: var(--accent);
  }

  ${({ fullWidth }) => fullWidth && 'width: 100%;'}
`;

const Input = styled.input`
  display: none;
`;

const I = styled.i`
  word-break: break-word;
`;

export function InputFile({
  name,
  label,
  error,
  passRef,
  id = name,
  fullWidth,
  accept,
  children
}) {
  const [fileName, setFileName] = useState();

  return (
    <Wrapper>
      <Label htmlFor={id} error={error}>
        {error || label} {fileName && <I>{fileName}</I>}
      </Label>
      <StyledInputFile fullWidth={fullWidth}>
        <Input
          type="file"
          ref={passRef}
          id={id}
          name={name}
          accept={accept}
          onChange={(e) => setFileName(e.currentTarget.files[0]?.name)}
        />
        {children}
      </StyledInputFile>
    </Wrapper>
  );
}
