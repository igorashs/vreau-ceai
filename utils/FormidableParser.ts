import * as validator from '@/utils/validator';
import FormidableNameSpace, { Files, File, IncomingForm } from 'formidable';
import { NextApiRequest } from 'next';
import { Formidable, ProductFields } from 'types';
import { productMessages } from './validator/schemas/product';
import { NextApiSessionRequest } from './withSession';

type ParsedData = {
  fields: Partial<ProductFields>;
  files: Files & { src?: File };
};

type HandleProductForm = (
  req: NextApiRequest | NextApiSessionRequest,
  options: Partial<FormidableNameSpace.Options>,
  allowedFileTypes: string[],
) => Promise<ParsedData>;

const handleProductForm: HandleProductForm = async (
  req,
  options,
  allowedFileTypes,
) => {
  const data: ParsedData = await new Promise((resolve, reject) => {
    const form: Formidable = new IncomingForm(options);

    form.onPart = (part) => {
      if (part.filename) {
        // get file extension
        const fileType = part.mime?.split('/').pop() ?? '';

        // validate allowed file types
        if (!allowedFileTypes.includes(fileType)) {
          form.emit(
            'error',
            validator.createValidationError({
              message: productMessages.src.invalid,
              key: 'src',
            }),
          );
        }
      }

      form.handlePart(part);
    };

    form.parse(req, (err, fields, files) => {
      if (err) {
        // custom error on maxFileSize
        if (err.message.includes('maxFileSize')) {
          reject(
            validator.createValidationError({
              message: productMessages.src.max,
              key: 'src',
            }),
          );
        } else {
          reject(err);
        }
      }

      resolve({ fields, files });
    });
  });

  return data;
};

export default {
  handleProductForm,
};
