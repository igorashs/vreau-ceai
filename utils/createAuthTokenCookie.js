import jwt from 'jsonwebtoken';
import cookie from 'cookie';

const EXPIRE_TIME = 24 * 60 * 60;

export const createAuthTokenCookie = (payload) => {
  const authToken = jwt.sign(payload, process.env.AUTH_PRIVATE_KEY, {
    expiresIn: EXPIRE_TIME
  });

  return cookie.serialize('auth-token', authToken, {
    secure: process.env.NODE_ENV !== 'development',
    httpOnly: true,
    sameSite: 'strict',
    path: '/',
    maxAge: EXPIRE_TIME
  });
};
