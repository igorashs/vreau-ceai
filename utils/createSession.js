import jwt from 'jsonwebtoken';
import cookie from 'cookie';

const EXPIRE_TIME = 24 * 60 * 60;

export const createSession = (payload) => {
  const token = jwt.sign(payload, process.env.AUTH_PRIVATE_KEY, {
    expiresIn: EXPIRE_TIME
  });

  const [header, jwtPayload, signature] = token.split('.');
  const session = header + '.' + jwtPayload;

  const sessionCookie = cookie.serialize('session', session, {
    secure: process.env.NODE_ENV !== 'development',
    sameSite: 'strict',
    path: '/',
    maxAge: EXPIRE_TIME
  });

  const authCookie = cookie.serialize('auth', signature, {
    secure: process.env.NODE_ENV !== 'development',
    httpOnly: true,
    sameSite: 'strict',
    path: '/',
    maxAge: EXPIRE_TIME
  });

  return [sessionCookie, authCookie];
};
