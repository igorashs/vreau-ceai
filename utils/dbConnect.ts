import mongoose from 'mongoose';

const dbConnect = async () => {
  if (mongoose.connection.readyState >= 1) {
    return;
  }

  await mongoose.connect(process.env.MONGODB_URI as string, {
    useNewUrlParser: true,
    useUnifiedTopology: true,
    useCreateIndex: true,
    useFindAndModify: false,
  });
};

export default dbConnect;
