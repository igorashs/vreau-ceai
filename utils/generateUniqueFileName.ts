import { nanoid } from 'nanoid';

const generateUniqueFileName = (fileName: string) => {
  // grab file extension
  const matches = fileName?.match(/\..+$/);

  // create new file name
  const uniqueFileName = nanoid() + (matches && matches[0]);

  return uniqueFileName;
};

export default generateUniqueFileName;
