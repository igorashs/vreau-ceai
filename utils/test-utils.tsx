import { render, RenderOptions } from '@testing-library/react';
import GlobalStyle from 'GlobalStyle';

const Wrapper: React.FC = ({ children }) => (
  <>
    <GlobalStyle />
    {children}
  </>
);

const customRender = (
  ui: React.ReactElement,
  options?: Omit<RenderOptions, 'queries'>,
) => render(ui, { wrapper: Wrapper, ...options });

export * from '@testing-library/react';

export { customRender as render };
