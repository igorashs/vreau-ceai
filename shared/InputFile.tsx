import styled from 'styled-components';
import { Label } from '@/shared/Label';

const Wrapper = styled.div`
  display: flex;
  flex-direction: column;
  gap: 7px;
`;

interface StyledInputFileProps {
  fullWidth?: boolean;
}

const StyledInputFile = styled.label<StyledInputFileProps>`
  align-self: flex-start;
  padding: calc(var(--baseline) / 4) calc(var(--baseline) / 2);
  border-radius: 4px;
  border: 0;
  color: var(--text-light);
  background-color: var(--accent-dark);
  box-shadow: 1px 1px 2px #00000066;
  text-transform: lowercase;
  text-decoration: none;

  cursor: pointer;

  :hover {
    background-color: var(--accent);
  }

  ${({ fullWidth }) => fullWidth && 'width: 100%;'}
`;

const Input = styled.input`
  display: none;
`;

interface InputFileProps {
  name: string;
  label: string;
  error?: string;
  passRef?:
    | ((instance: HTMLInputElement | null) => void)
    | React.RefObject<HTMLInputElement>;
  id?: string;
  accept?: string;
  fullWidth?: boolean;
}

export const InputFile = ({
  name,
  label,
  error,
  passRef,
  id = name,
  fullWidth,
  accept,
  children,
}: React.PropsWithChildren<InputFileProps>) => {
  return (
    <Wrapper>
      <Label htmlFor={id} error={!!error}>
        {error || label}
      </Label>
      <StyledInputFile fullWidth={fullWidth}>
        <Input type="file" ref={passRef} id={id} name={name} accept={accept} />
        {children}
      </StyledInputFile>
    </Wrapper>
  );
};
