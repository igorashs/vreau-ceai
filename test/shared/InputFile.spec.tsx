import { render } from '@/utils/test-utils';
import InputFile from '@/shared/InputFile';

describe('InputFile', () => {
  it('renders correctly', () => {
    const name = 'file';
    const label = 'file label';

    const { container, getByText, getByLabelText } = render(
      <InputFile name={name} label={label}>
        content
      </InputFile>,
    );

    const input = getByLabelText('content');

    expect(input).toHaveAttribute('name', name);
    expect(input).toHaveAttribute('id', name);
    expect(input).toHaveAttribute('type', 'file');
    expect(getByText('content')).toBeVisible();
    expect(getByText(label)).toBeVisible();
    expect(container.firstChild).toMatchSnapshot();
  });

  describe('error prop', () => {
    it('renders with error message', () => {
      const name = 'file';
      const label = 'file label';
      const errorMsg = 'error';

      const { queryByText, getByText } = render(
        <InputFile name={name} label={label} error={errorMsg}>
          content
        </InputFile>,
      );

      expect(queryByText(label)).not.toBeInTheDocument();
      expect(getByText(errorMsg)).toBeVisible();
    });
  });

  describe('fullWidth prop', () => {
    it('renders with fulWidth', () => {
      const name = 'file';
      const label = 'file label';

      const { getByLabelText } = render(
        <InputFile name={name} label={label} fullWidth>
          content
        </InputFile>,
      );

      expect(getByLabelText('content').parentElement).toHaveStyleRule(
        'width',
        '100%',
      );
    });
  });
});
