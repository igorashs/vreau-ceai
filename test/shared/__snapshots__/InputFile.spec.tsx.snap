// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`InputFile renders correctly 1`] = `
.c1 {
  font-weight: 500;
}

.c0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  gap: 7px;
}

.c2 {
  -webkit-align-self: flex-start;
  -ms-flex-item-align: start;
  align-self: flex-start;
  padding: calc(var(--baseline) / 4) calc(var(--baseline) / 2);
  border-radius: 4px;
  border: 0;
  color: var(--text-light);
  background-color: var(--accent-dark);
  box-shadow: 1px 1px 2px #00000066;
  text-transform: lowercase;
  -webkit-text-decoration: none;
  text-decoration: none;
  cursor: pointer;
}

.c2:hover {
  background-color: var(--accent);
}

.c3 {
  display: none;
}

<div
  class="c0"
>
  <label
    class="c1"
    for="file"
  >
    file label
  </label>
  <label
    class="c2"
  >
    <input
      class="c3"
      id="file"
      name="file"
      type="file"
    />
    content
  </label>
</div>
`;
